lib/matplotlib/figure.py

```diff
@@ -1462,67 +1518,22 @@ def legend(self, *args, **kwargs):
-        if len(args) == 0:
-            handles = []
-            labels = []
-
-            def in_handles(h, l):
-                # Method to check if we already have a given handle and label.
-                # Consider two handles to be the same if they share a label,
-                # color, facecolor, and edgecolor.
-
-                # Loop through each handle and label already collected
-                for f_h, f_l in zip(handles, labels):
-                    if f_l != l:
-                        continue
-                    if type(f_h) != type(h):
-                        continue
-                    try:
-                        if f_h.get_color() != h.get_color():
-                            continue
-                    except AttributeError:
-                        pass
-                    try:
-                        if f_h.get_facecolor() != h.get_facecolor():
-                            continue
-                    except AttributeError:
-                        pass
-                    try:
-                        if f_h.get_edgecolor() != h.get_edgecolor():
-                            continue
-                    except AttributeError:
-                        pass
-                    return True
-                return False
-
-            for ax in self.axes:
-                ax_handles, ax_labels = ax.get_legend_handles_labels()
-                for h, l in zip(ax_handles, ax_labels):
-                    if not in_handles(h, l):
-                        handles.append(h)
-                        labels.append(l)
-            if len(handles) == 0:
-                warnings.warn("No labeled objects found. "
-                              "Use label='...' kwarg on individual plots.")
-                return None
-
-        elif len(args) == 2:
-            # LINES, LABELS
-            handles, labels = args
-
-        elif len(args) == 3:
-            # LINES, LABELS, LOC
-            handles, labels, loc = args
-            kwargs['loc'] = loc
-
-        else:
-            raise TypeError('Invalid number of arguments passed to legend. '
-                            'Please specify either 0 args, 2 args '
-                            '(artist handles, figure labels) or 3 args '
-                            '(artist handles, figure labels, legend location)')
-
-        l = Legend(self, handles, labels, **kwargs)
+        handles, labels, extra_args, kwargs = mlegend._parse_legend_args(
+                self.axes,
+                *args,
+                **kwargs)
+        # check for third arg
+        if len(extra_args):
+            # cbook.warn_deprecated( 
+            #     "2.1",
+            #     "Figure.legend will accept no more than two "
+            #     "positional arguments in the future.  Use "
+            #     "'fig.legend(handles, labels, loc=location)' "
+            #     "instead.")
+            # kwargs['loc'] = extra_args[0]
+            # extra_args = extra_args[1:]
+            pass
```

Comment:
@tacaswell  This now fails Travis - what was the plan here?  I thought we were keeping `figure.legend` w/ three possible args?

Reply:
This shouldn't have changed anything other than to prevent `fig.legend(a, b, l, loc=c)` case.

Reply:
Ah, I see:

```
self = <matplotlib.tests.test_legend.TestLegendFigureFunction object at 0x7fffd6b1fc50>
    def test_legend_label_three_args(self):
        fig, ax = plt.subplots()
        lines = ax.plot(range(10))
        with mock.patch('matplotlib.legend.Legend') as Legend:
            fig.legend(lines, ['foobar'], 'right')
>       Legend.assert_called_with(fig, lines, ['foobar'], loc='right')
```

The test was checking that `loc` got loaded w/ the third element.  I'll fix later. Thanks!

Reply:
@tacaswell  Changed `legend_test` to check for an Exception should someone call: `fig.legend(handles, labels, 'right', loc='left')`  (`test_legend_label_three_args_pluskw`)  I think thats what you wanted.  

Any reason to leave this check/pass in here?  Just posterity sake?

Reply:
To avoid changing the API as much as possible.  I am :+1: on deprecating the other positional arguments for 2.2, but not in 2.1.1 (so leave that code in place so we can just un-comment it in a follow on PR).

I don't think we needed a test for what is core-python behavior (you can not pass a value as both positional and keyword) nor should we break it!